README.md
[![pub package](https://img.shields.io/pub/v/decorated_icon.svg)](https://pub.dev/packages/decorated_icon)

# Decorated Icon
The best-performing shadowed icon widget for Flutter!

`DecoratedIcon` is a super lightweight `Icon` that supports shadow decorations.

No complicated hacks or multiple widgets, just `List<Shadow> shadows` like a `TextStyle`!

This widget is able to achieve the most efficient implementation of an icon with a shadow because it is uses a single `Text` widget to display the icon and apply the shadows.

It even supports debugging and `Semantics`.

One thing to keep in mind when using multiple shadows is the order of stacking by Flutter: the first shadow in the list will be painted at the bottom. (Similar to how the [Stack](https://api.flutter.dev/flutter/widgets/Stack-class.html) widget draws its children.)

Also note that due to Flutter's current limitation, achieving an **"outer glow"** effect is not currently possible. You can read more [here](https://github.com/flutter/flutter/issues/65763).


## Usage
To use this package, add `decorated_icon` as a [dependency in your pubspec.yaml file](https://flutter.io/using-packages/).

## Example
![snapshot](example/assets/images/snapshot.png)

Import the library.

``` dart
import 'package:decorated_icon/decorated_icon.dart';
```

Then use `DecoratedIcon` as you would an `Icon`:
```dart
final decoratedIcons = Row(
  mainAxisAlignment: MainAxisAlignment.spaceEvenly,
  children: const [
    DecoratedIcon(
      Icons.android,
      color: Colors.purple,
      size: 60.0,
      shadows: [
        BoxShadow(
          blurRadius: 42.0,
          color: Colors.purpleAccent,
        ),
        BoxShadow(
          blurRadius: 12.0,
          color: Colors.white,
        ),
      ],
    ),
    DecoratedIcon(
      Icons.favorite,
      color: Colors.lightBlue.shade50,
      size: 60.0,
      shadows: [
        BoxShadow(
          blurRadius: 12.0,
          color: Colors.blue,
        ),
        BoxShadow(
          blurRadius: 12.0,
          color: Colors.green,
          offset: Offset(0, 6.0),
        ),
      ],
    ),
    DecoratedIcon(
      Icons.fingerprint,
      color: Colors.orange,
      size: 60.0,
      shadows: [
        BoxShadow(
          color: Colors.black,
          offset: Offset(3.0, 3.0),
        ),
      ],
    ),
  ],
);
```


CHANGELOG.md
## [1.2.0]

- Added support for TextDirection, Semantics, and Debugging.
- Minor updates in README.md.

## [1.1.0]

- Migrated to Null safety.
- Fixed spelling mistake in README.md file.

## [1.0.3]

- Added `icon.fontPackage` to the icon's text style.

## [1.0.2]

- Updated description.

## [1.0.1]

- Updated description with more information.

## [1.0.0]

- Initial release.

lib/decorated_icon.dart
/// Provides a [DecoratedIcon] that can be easily decorated with shadows.
library decorated_icon;

import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart'
    show DiagnosticPropertiesBuilder, DoubleProperty;

/// An icon that can be easily decorated with shadows.
class DecoratedIcon extends StatelessWidget {
  /// An icon that can be easily decorated with shadows.
  const DecoratedIcon(
    this.icon, {
    Key? key,
    this.size,
    this.color,
    this.shadows,
    this.semanticLabel,
    this.textDirection,
  }) : super(key: key);

  /// The color of the icon.
  ///
  /// Defaults to white, based on defaults for uninherited [TextStyle] values.
  final Color? color;

  /// The [IconData] to use as the icon. (E.g.: `Icons.favorite`)
  final IconData icon;

  /// A [List] of [Shadow] properties to apply to the icon.
  ///
  /// Note that Flutter stacks shadows in reverse order.
  /// Meaning the bottom-most shadow will display over the other shadows in the list.
  final List<Shadow>? shadows;

  /// The size of the icon in logical pixels.
  ///
  /// Defaults to `10.0`, based on defaults for uninherited [TextStyle] values.
  final double? size;

  /// Label announced with accessibility modes active
  /// (such as TalkBack or VoiceOver) but not shown in the UI.
  final String? semanticLabel;

  /// Will have no effect if [icon]'s [IconData.matchTextDirection] field is `false`,
  /// but a `DecoratedIcon` must provide text direction for specification coherence:
  /// either directly with this value or, if `null`, ambient [Directionality.of].
  final TextDirection? textDirection;

  @override
  Widget build(BuildContext context) {
    final iconDirection = textDirection ?? Directionality.of(context);

    Widget iconWidget = Text(
      String.fromCharCode(icon.codePoint),
      style: TextStyle(
        color: color,
        decoration: TextDecoration.none,
        fontFamily: icon.fontFamily,
        fontWeight: FontWeight.normal,
        fontSize: size,
        height: 1,
        inherit: false,
        package: icon.fontPackage,
        shadows: shadows,
      ),
    );

    if (icon.matchTextDirection) {
      switch (iconDirection) {
        case TextDirection.rtl:
          // Mirrors over Y-axis
          iconWidget = Transform(
            transform: Matrix4.identity()..scale(-1.0, 1.0, 1.0), // scales X
            alignment: Alignment.center,
            transformHitTests: false,
            child: iconWidget,
          );
          break;
        case TextDirection.ltr:
          break;
      }
    }

    return Semantics(
      label: semanticLabel,
      child: ExcludeSemantics(child: iconWidget),
    );
  }

  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
    super.debugFillProperties(properties);
    properties.add(
        IconDataProperty('icon', icon, ifNull: '<empty>', showName: false));
    properties.add(DoubleProperty('size', size, defaultValue: null));
    properties.add(ColorProperty('color', color, defaultValue: null));
  }
}


example/lib/main.dart
import 'package:decorated_icon/decorated_icon.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'DecoratedIcon Example',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  MyHomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceEvenly,
          children: [
            DecoratedIcon(
              Icons.android,
              color: Colors.purple,
              size: 60.0,
              shadows: [
                BoxShadow(
                  blurRadius: 42.0,
                  color: Colors.purpleAccent,
                ),
                BoxShadow(
                  blurRadius: 12.0,
                  color: Colors.white,
                ),
              ],
            ),
            DecoratedIcon(
              Icons.favorite,
              color: Colors.lightBlue.shade50,
              size: 60.0,
              shadows: [
                BoxShadow(
                  blurRadius: 12.0,
                  color: Colors.blue,
                ),
                BoxShadow(
                  blurRadius: 12.0,
                  color: Colors.green,
                  offset: Offset(0, 6.0),
                  spreadRadius: 60.0,
                ),
              ],
            ),
            DecoratedIcon(
              Icons.fingerprint,
              color: Colors.orange,
              size: 60.0,
              shadows: [
                BoxShadow(
                  color: Colors.black,
                  offset: Offset(3.0, 3.0),
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}
